python/GafferSceneUI/ShaderTweaksUI.py

```diff
@@ -279,63 +279,35 @@ def __menuDefinition( self ) :
 
 		return result
 
-	def __addFromAffectedMenuDefinition( self ) :
+	def __browseAffectedShaders( self ) :
 
-		return self.__addFromPathsMenuDefinition( _pathsFromAffected( self ) )
+		self.__browseShaders( _pathsFromAffected( self ), "Affected Shaders" )
 
-	def __addFromSelectedMenuDefinition( self ) :
+	def __browseSelectedShaders( self ) :
 
-		return self.__addFromPathsMenuDefinition( _pathsFromSelection( self ) )
+		self.__browseShaders( _pathsFromSelection( self ), "Selected Shaders" )
 
-	def __addFromPathsMenuDefinition( self, paths ) :
-
-		result = IECore.MenuDefinition()
+	def __browseShaders( self, paths, title ) :
 
 		shaderAttributes = _shaderAttributes( self, paths, affectedOnly = True )
-		if not len( shaderAttributes ) :
-			result.append(
-				"/No Shaders Found", { "active" : False }
-			)
-			return result
-
-		shaders = {}
-		for attributes in shaderAttributes.values() :
-			for attributeName, network in attributes.items() :
-				for shaderName, shader in network.shaders().items() :
-					if shaderName == network.getOutput().shader :
-						shaderName = ""
-					shaderParameters = shaders.setdefault( shaderName, {} )
-					for parameterName, parameterValue in shader.parameters.items() :
-						if parameterName.startswith( "__" ) :
-							continue
-						shaderParameters[parameterName] = parameterValue
-
-		if not len( shaders ) :
-			result.append(
-				"/No Parameters Found", { "active" : False }
-			)
-			return result
-
-		for shaderName, shader in shaders.items() :
-
-			menuPrefix = "/"
-			tweakPrefix = ""
-			if len( shaders ) > 1 :
-				menuPrefix = "/Other/{0}/".format( shaderName ) if shaderName else "/Main/"
-				tweakPrefix = "{0}.".format( shaderName ) if shaderName else ""
-
-			for parameterName in sorted( shader.keys() ) :
-				result.append(
-					menuPrefix + parameterName,
-					{
-						"command" :	functools.partial(
-							Gaffer.WeakMethod( self.__addTweak ),
-							tweakPrefix + parameterName, shader[parameterName]
+
+		shaderNetworks = set().union( *[ set( a.values() ) for a in shaderAttributes.values() ] )
+
+		parentWindow = self.ancestor( GafferUI.ScriptWindow )
+		browser = GafferSceneUI.ShaderUI._ShaderParameterDialogue( shaderNetworks, title, parentWindow )
+
+		tweaks = browser.waitForParameters( parentWindow = parentWindow )
+
+		if tweaks is not None :
+			for shaderName, parameter in tweaks :
+				for network in shaderNetworks :
+					if shaderName in network.shaders() and parameter in network.shaders()[shaderName].parameters :
+						self.__addTweak(
+							shaderName + "." + parameter,
+							network.shaders()[shaderName].parameters[parameter]
 						)
-					}
-				)
 
-		return result
+						break
```

Comment:
I haven't figured out why, but this doesn't seem to be doing anything when I choose a parameter from the light shader in the example below :

<details>

```
import Gaffer
import GafferArnold
import GafferScene
import IECore
import imath

Gaffer.Metadata.registerValue( parent, "serialiser:milestoneVersion", 0, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:majorVersion", 61, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:minorVersion", 8, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["SpotLight"] = GafferArnold.ArnoldLight( "SpotLight" )
parent.addChild( __children["SpotLight"] )
__children["SpotLight"].loadShader( "spot_light" )
__children["SpotLight"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ShaderTweaks"] = GafferScene.ShaderTweaks( "ShaderTweaks" )
parent.addChild( __children["ShaderTweaks"] )
__children["ShaderTweaks"]["tweaks"].addChild( GafferScene.TweakPlug( Gaffer.BoolPlug( "value", defaultValue = True, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), "__shader_cast_shadows", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ShaderTweaks"]["tweaks"].addChild( GafferScene.TweakPlug( Gaffer.FloatPlug( "value", defaultValue = 1.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), "__shader_aspect_ratio", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ShaderTweaks"]["tweaks"].addChild( GafferScene.TweakPlug( Gaffer.FloatPlug( "value", defaultValue = 1.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), "__shader_aspect_ratio1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ShaderTweaks"]["tweaks"].addChild( GafferScene.TweakPlug( Gaffer.FloatPlug( "value", defaultValue = 1.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), "__shader_diffuse", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ShaderTweaks"]["tweaks"].addChild( GafferScene.TweakPlug( Gaffer.FloatPlug( "value", defaultValue = 1.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), "__shader_diffuse1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ShaderTweaks"]["tweaks"].addChild( GafferScene.TweakPlug( Gaffer.FloatPlug( "value", defaultValue = 1.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), "__shader_aspect_ratio2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ShaderTweaks"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["PathFilter"] = GafferScene.PathFilter( "PathFilter" )
parent.addChild( __children["PathFilter"] )
__children["PathFilter"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["SpotLight"]["name"].setValue( 'spotLight' )
__children["SpotLight"]["__uiPosition"].setValue( imath.V2f( 24.0635796, -13.138607 ) )
__children["ShaderTweaks"]["in"].setInput( __children["SpotLight"]["out"] )
__children["ShaderTweaks"]["filter"].setInput( __children["PathFilter"]["out"] )
__children["ShaderTweaks"]["shader"].setValue( 'ai:light' )
__children["ShaderTweaks"]["tweaks"]["__shader_cast_shadows"]["name"].setValue( '__shader.cast_shadows' )
__children["ShaderTweaks"]["tweaks"]["__shader_aspect_ratio"]["name"].setValue( '__shader.aspect_ratio' )
__children["ShaderTweaks"]["tweaks"]["__shader_aspect_ratio1"]["name"].setValue( '__shader.aspect_ratio' )
__children["ShaderTweaks"]["tweaks"]["__shader_diffuse"]["name"].setValue( '__shader.diffuse' )
__children["ShaderTweaks"]["tweaks"]["__shader_diffuse1"]["name"].setValue( '__shader.diffuse' )
__children["ShaderTweaks"]["tweaks"]["__shader_aspect_ratio2"]["name"].setValue( '__shader.aspect_ratio' )
__children["ShaderTweaks"]["__uiPosition"].setValue( imath.V2f( 24.0635796, -21.3026695 ) )
__children["PathFilter"]["paths"].setValue( IECore.StringVectorData( [ '/spotLight' ] ) )
__children["PathFilter"]["__uiPosition"].setValue( imath.V2f( 37.2302475, -15.2206287 ) )


del __children

```
</details>

Reply:
I forget what we discussed doing about using `ShaderName.parameterName` or just `parameterName` when the parameter is on the output shader, but right now I'm thinking it'd be desirable to keep the previous behaviour and just use `parameterName` in this case. Perhaps that's a bad idea though?

Reply:
My worry with using just `parameterName` is that now that we are mixing shaders, you could have `shaderA` which is the output of a network (and can use `parameterName`) and elsewhere `shaderA` which is not an output (and requires `shaderA.parameterName`).

Perhaps that's something the client code - `ShaderTweaks` and related - should be expected to handle? And perhaps it's a rare case in reality, and `ShaderTweaks` could handle it correctly by adding two tweaks in the rare instances it happens?

Reply:
Yeah, fair point, and there's definitely a benefit to using the same pattern everywhere. I think what concerns me is that light tweaks are a very common use case, and in many (the majority of?) cases they only contain a single shader, and it has a horrible `__shader` handle. Maybe I'd feel better about it if we could just sort the handle out (in which case the worry is code or files that are relying on it being `__shader`).

Reply:
After discussin offline, we're now dropping the `shader.` prefix if _all_ of the shaders for that parameter are output shaders. That is done in 695714a4f121f430e45d5cb098c841426429d6a2, which has the side effect of fixing the light parameter problem above.

Reply:
And the `__shader` handle issue is fixed in 695714a4f121f430e45d5cb098c841426429d6a2 where we replace `__shader` with `Lights` in the display only.

I didn't make the test conditional on shader type as would be ideal because for multiple lights of different types, the `type` property is `---` which will fail the test. If this is a weak test, I can find a better solution, but going for simplicity for now.